### PR TITLE
Closes #5715: Login storage refactor

### LINKED ISSUE #5715: SyncWorker: setLogins - Tried to use a null handle
Sometimes the Fenix app gets into a Handle invalid state. In my experience I think it was when I worked with local servers with servers sending errors.

Logcat reports the following:

 
Handle is invalid: Tried to use a null handle (this object has probably been closed)

```
Work [ id=113443b2-a824-415e-a1cf-11f55b0383b9, tags={ mozilla.components.service.fxa.sync.WorkManagerSyncWorker, Debounce, Common } ] failed because it threw an exception/error
                         E  java.util.concurrent.ExecutionException: mozilla.appservices.syncmanager.UnexpectedError: Handle is invalid: Tried to use a null handle (this object has probably been closed)
                         E      at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:516)
                         E      at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:475)
                         E      at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:284)
                         E      at androidx.work.impl.utils.SerialExecutor$Task.run(SerialExecutor.java:91)
                         E      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
                         E      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
                         E      at java.lang.Thread.run(Thread.java:919)
                         E  Caused by: mozilla.appservices.syncmanager.UnexpectedError: Handle is invalid: Tried to use a null handle (this object has probably been closed)
                         E      at mozilla.appservices.syncmanager.RustError.intoException(RustError.kt:40)
                         E      at mozilla.appservices.syncmanager.SyncManager.setLogins(SyncManager.kt:96)
                         E      at mozilla.components.service.fxa.sync.WorkManagerSyncWorker.doSync(WorkManagerSyncManager.kt:342)
                         E      at mozilla.components.service.fxa.sync.WorkManagerSyncWorker$doWork$2.invokeSuspend(WorkManagerSyncManager.kt:326)
                         E      at mozilla.components.service.fxa.sync.WorkManagerSyncWorker$doWork$2.invoke(Unknown Source:10)
                         E      at mozilla.components.service.sync.logins.SyncableLoginsStore.unlocked(AsyncLoginsStorage.kt:466)
                         E      at mozilla.components.service.sync.logins.SyncableLoginsStore$unlocked$1.invokeSuspend(Unknown Source:12)
                         E      at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                         E      at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:561)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:727)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:667)
                         E      at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:655)
                         I  Worker result FAILURE for Work [ id=113443b2-a824-415e-a1cf-11f55b0383b9, tags={ mozilla.components.service.fxa.sync.WorkManagerSyncWorker, Debounce, Common } ]

```

This happens when there are 401 errors from the sync server

### PR DESCRIPTION
**Pull request by grigoryk**
---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by liuche** [comment]
@csadilek when this lands, can you make an AC dot release at the same time? We'll need one for 4.0 anyway.

File: components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+            }
+        }
+    }
+
+    internal fun getStorage(): RustLoginStorage = synchronized(this) {
+        check(storage != null) { "must call init first" }
+        return storage!!
+    }
+
+    internal fun getHandle(): Long {
```
Comment (csadilek) [inline components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt]: Looks like this should be `synchronized` as well?

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt [hunk 2]
```diff
@@ -5,6 +5,99 @@
...
 import kotlinx.coroutines.Deferred
+import org.json.JSONObject
+
+/**
+ * An interface describing a storage layer for logins/passwords.
+ */
+@SuppressWarnings("TooManyFunctions")
+interface LoginsStorage : AutoCloseable {
+    /**
+     * Delete all login records. These deletions will be synced to the server on the next call to sync.
```
Comment (csadilek) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt]: nit: Deletes...

File: components/concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt [hunk 3]
```diff
@@ -5,6 +5,99 @@
...
+
+    /**
+     * Clear out all local state, bringing us back to the state before the first write (or sync).
+     */
+    suspend fun wipeLocal()
+
+    /**
+     * Deletes the password with the given ID.
+     *
+     * Resolves to true if the deletion did anything.
```
Comment (csadilek) [inline components/concept/storage/src/main/java/mozilla/components/concept/storage/LoginsStorage.kt]: nit: Add `@param`, `@return`, for all of them.

**Comment by grigoryk** [comment]
bors r=csadilek

**Pull request merged**